source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -106,33 +106,44 @@ DnsCacheImpl::loadDnsCacheEntry(absl::string_view raw_host, uint16_t default_por
             is_proxy_lookup ? "proxy mode " : "");
   ThreadLocalHostInfo& tls_host_info = *tls_slot_;
 
-  auto [is_overflow, host_info] = [&]() {
+  bool is_overflow = false;
+  absl::optional<DnsHostInfoSharedPtr> host_info = absl::nullopt;
+  bool ignore_cached_entries = false;
+
+  {
     absl::ReaderMutexLock read_lock{&primary_hosts_lock_};
+    is_overflow = primary_hosts_.size() >= max_hosts_;
     auto tls_host = primary_hosts_.find(host);
-    return std::make_tuple(
-        primary_hosts_.size() >= max_hosts_,
-        (tls_host != primary_hosts_.end() && tls_host->second->host_info_->firstResolveComplete())
-            ? absl::optional<DnsHostInfoSharedPtr>(tls_host->second->host_info_)
-            : absl::nullopt);
-  }();
+    if (tls_host != primary_hosts_.end() && tls_host->second->host_info_->firstResolveComplete()) {
+      host_info = tls_host->second->host_info_;
+    }
+  };
 
   if (host_info) {
     ENVOY_LOG(debug, "cache hit for host '{}'", host);
-    return {LoadDnsCacheEntryStatus::InCache, nullptr, host_info};
-  } else if (is_overflow) {
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.reresolve_null_addresses") &&
+        !is_proxy_lookup && *host_info && (*host_info)->address() == nullptr) {
+      ENVOY_LOG(debug, "ignoring null address cache hit for miss for host '{}'", host);
+      ignore_cached_entries = true;
```

Comment:
instead of using the variable `ignore_cached_entries`, couldn't we just call `removeHost()` here, and then fall through to the logic on L139?

Reply:
that way, we wouldn't need the extra param to `startCacheLoad()`

Reply:
we can't call removeHost inline because this is a worker thread function and that's a main thread function.  So we can either
dispatcher.post(removeHost)
dispatcher.post(startCacheLoad())
or dispatcher.post(startCacheLoad(remove_host = true)).
I don't think it makes much of a perf difference - I found one context switch mildly easier to follow but I'm happy to remove the arg and do 2 calls if you think it's cleaner.

Reply:
ah ok, thanks for explaining, SGTM